Functions/Gherkin.ps1

```diff
@@ -787,3 +787,150 @@ function ConvertTo-HashTableArray {
     }
 }
 
+function Import-FeatureFile {
+    param(
+        # Path to Search
+        [Parameter(ValueFromPipeline = $true)]
+        [string[]]
+        $Path = $pwd
+    )
+    
+    begin {
+        $parser = New-Object Gherkin.Parser
+    }
+
+    process {
+        $item = Get-Item -Path $Path
+
+        if ($item.psisContainer) {
+            foreach ($file in $path | Get-ChildItem -Recurse -Filter *.feature) {
+                try {
+                    $fixtureFilename = "{0}.Steps.ps1" -f $file.BaseName
+                    $fullName = Join-Path -Path $file.Directory -ChildPath $fixtureFilename
+                    $parsed = $parser.parse($file.FullName)
+
+                    if (Test-Path -Path $fullName) {
+                        Update-TestStubs -stubFile $fullName -parsedInput $parsed
+                    }
+                    else {
+                        $stubFile = New-Item -Path $fullName -ItemType File
+                        
+                        Add-Stubs -stubFile $stubFile -parsedInput $parsed   
+                    }
+                }
+                catch {
+                    throw $_
+                }
+            }   
+        }
+        else {
+            try {
+                $fixtureFilename = "{0}.Steps.ps1" -f $item.BaseName
+                $fullName = Join-Path -Path $item.Directory -ChildPath $fixtureFilename
+                $parsed = $parser.parse($item.FullName)
+
+                if (Test-Path -Path $fullName) {
+                    Update-TestStubs -stubFile $fullName -parsedInput $parsed
+                }
+                else {
+                    $stubFile = New-Item -Path $item.Directory -Name $fixtureFilename -ItemType File
+                    
+                    Add-Stubs -stubFile $stubFile -parsedInput $parsed
+                    
+                }
+            }
+            catch {
+                throw $_
+            }
+        }
+    }
+
+    end {
+
+    }
+
+
+}
+
+function Add-Stubs {
```

Comment:
You would need to check and determines whether the stub already exists --in any .Steps file. There's a function `Find-GherkinStep` you could use. You shouldn't re-generate steps that already have a definition, that could cause ambiguity errors.

Reply:
If the user then writes two feature files that really do have scenarios by the same name, are we saying that we shouldn't allow that? It seems like that would be dictating to the user what their workflow should be.

Reply:
This is simply a namespace/scope issue. 

In a single application (or module), where you have all your specs (.feature files) together in one folder, you don't want `When I create a user` to have different implementations depending on which file it's in -- that would be confusing. 

That is: when you have two *steps* with the same text, they should match the _exact_ same implementation across all your features. In fact, I frequently create a single "Steps" file, regardless of the number of feature files, so I don't have to keep track of which steps I've already implemented in another file.

Reply:
I agree that it would be confusing to have a single scenario with multiple implementations, but my question would be, how much of a single view of how to organize one's work should be enforced by Pester? 

I personally had envisioned a very limited scope of behavior for this feature, which is that for each feature file the code finds, it should create a corresponding steps file with the scenario stubs implemented, such that the behavior is as predictable or as understandable after the fact as possible by reading the documentation for Cucumber, not for Pester.

With that in mind, I agree that multiple implementations could be confusing, but that confusion would be the responsibility of the developer to either resolve, or decide that the duplication is intentional. On the other hand, if that duplication is intentional, for whatever reason, and Pester refuses to implement the test stub, that confusion is now our fault.

So when we talk about tracking which stubs we have implemented and avoiding duplication, I think of some of the following scenarios:
1. Two different features files in the same directory specify the same scenario  
  Should we create only a single steps file for all features files in the directory? If that's the case then given that the test runs are matched based on regex then I think we would indeed only implement the stub once. But I honestly don't prefer to work this way. I prefer lots of small features and steps files to keep them short and to the point. Perhaps we could support a switch parameter option that would force all steps files in a directory to combined into a single file to support both of our work flows.  
2. Two different features files in different directories specify the same scenario.  
  Should we make any attempt to prevent duplication across directories? Perhaps if we do want to support that, instead of a switch param to combine in a single directory, we could support a ```-stepsScope``` param that would offer a choice of ```file```, ```directory```, ```all``` and if all is selected, a ```-stepsFileOutputPath``` param that would specify where the master steps file should be written, that would default to ```$pwd```

I bring up scenario number two because in the project I was asked to implement this feature for, we did indeed have the scenarios files spread out in the repository according to the place where it made the most sense to find them. Scenarios for server side code were in one folder, and scenarios that dictated how the client should behave were in a different folder that corresponded to the client.

Reply:
It's not about enforcing a single way of doing things. In particular, we cannot enforce any relationship between feature files and step files. That is, we don't load <FOO>.feature and look for <FOO>.steps.ps1, we load *.steps.ps1 recursively, and then we start running through features.

Part of the reason I like Gherkin is that I automatically get to re-use _step definitions_ in many of scenarios, which usually ends up meaning you write a lot less code than you typically do in Pester's RSpec-based syntax... 

Normally you keep all your .feature files **together**, because together they are the specification for your product. Obviously if you have one "project" with multiple products (i.e. multiple clients and a server), you can (and probably should) create separate specs folders...  Also, you can keep your .steps files in the same folder with the .feature files, or in a subfolder -- currently Pester's `Invoke-Gherkin` merely requires them to be next to or in the -Path you run Gherkin against.  Obviously in that case, you can _already_ run Invoke-Gherkin multiple times, against different folders, but then you don't get _one_ result count, and you don't get as much re-use.

## Step definitions are just functions. 

Pester's Gherkin (like all the others that I know of) finds *all* the steps at once, and basically sources them into Pester. If you define the same name in two different files, you're **overwriting** the definition, just as you would with functions in modules...

Imagine a user-provisioning module:

```gherkin
Feature: Creating AD User

Scenario: Simple
  Given we are an administrator
  When we create a user
  Then Get-ADUser should work
```

Imagine I've written a `ADUser.Steps.ps1` file and gotten everything working correctly.

Now it's time to add a new feature, so I add a feature file:

```gherkin
Feature: Creating O365 User

Scenario: Simple
  Given we are an administrator
  When we create a user
  Then Get-User should work
```

Then we generate stubs ... 

If you generate a new stub for `Given we are an administrator` and `When we create a user` in a new file `O365.Steps.ps1`, you _**will break** the existing tests_.  I don't see a way around that which doesn't force you to rewrite every step definition in every feature (which is certainly out of the question).

The bottom line is that we should avoid generating stubs that will break things. If you're running in a folder and the other step is defined in a different folder, you'll have no problems. If the other step is in the same folder, I'd suggest putting a comment in the generated file pointing out where the step is defined.

Reply:
Maybe then only generating a single file per folder would make it easier, both to ensure I don't create two stubs for the same function, and to drive that point home to the user.

Reply:
Might be! Because of the lack of tooling, it's _definitely_ easier to do that way right now. 

When I work with, say, SpecFlow in C#, they have good tooling in Visual Studio where I get intellisense while writing the .feature files to auto-complete step statements based on steps that I've already written. In that world, I don't care as much where my steps are, and I'm much more likely to break them up ... I believe they also have an indicator in the editor, while you're working on the feature file, to tell you if a step doesn't have an implementation ... and their tool for generating steps is also integrated to the editor.

In my ideal world, I'd love to write a Pester+Gherkin extension for VS Code, so I could run a _specific scenario_ from code, have the results show as red/green/black markers in the sidebar, like what https://wallabyjs.com/ does.  Such a plugin would be able to show you _as you write_ your feature file what the steps are that are already available, and would be able to generate new step placeholders ... all by calling the functions from Pester...

Reply:
As I read your description, an extension for VSCode is exactly where I was thinking that belonged too. I think I have a path forward now for this feature though at least. 

I'm switching jobs this week, so things are going to be chaotic for me for a little, but I'll start work as soon as I can.